core/src/main/java/gov/nist/secauto/metaschema/core/model/constraint/LoggingConstraintValidationHandler.java

```diff
@@ -191,59 +252,117 @@ public void handleExpectViolation(
         .map(IConstraint::getLevel)
         .max(Comparator.comparing(Level::ordinal))
         .get());
     if (isLogged(level)) {
-      logConstraint(level, target, newAllowedValuesViolationMessage(failedConstraints, target), null);
+      logConstraint(
+          level,
+          target,
+          newAllowedValuesViolationMessage(
+              failedConstraints,
+              target),
+          null);
     }
   }
 
   @Override
-  public void handleIndexDuplicateViolation(IIndexConstraint constraint, INodeItem node) {
+  public void handleIndexDuplicateViolation(
+      @NonNull IIndexConstraint constraint,
+      @NonNull INodeItem node,
+      @NonNull DynamicContext dynamicContext) {
+    // always log at level critical
     Level level = Level.CRITICAL;
     if (isLogged(level)) {
-      logConstraint(level, node, newIndexDuplicateViolationMessage(constraint, node), null);
+      logConstraint(
+          level,
+          node,
+          newIndexDuplicateViolationMessage(
+              constraint,
+              node),
+          null);
     }
   }
 
   @Override
-  public void handleIndexMiss(IIndexHasKeyConstraint constraint, INodeItem node, INodeItem target, List<String> key) {
+  public void handleIndexMiss(
+      @NonNull IIndexHasKeyConstraint constraint,
+      @NonNull INodeItem node,
+      @NonNull INodeItem target,
+      @NonNull List<String> key,
+      @NonNull DynamicContext dynamicContext) {
     Level level = constraint.getLevel();
     if (isLogged(level)) {
-      logConstraint(level, node, newIndexMissMessage(constraint, node, target, key), null);
+      logConstraint(
+          level,
+          node,
+          newIndexMissMessage(
+              constraint,
+              node,
+              target,
+              key,
+              dynamicContext),
+          null);
     }
   }
 
   @Override
-  public void handleMissingIndexViolation(IIndexHasKeyConstraint constraint, INodeItem node, INodeItem target,
-      String message) {
+  public void handleMissingIndexViolation(
+      @NonNull IIndexHasKeyConstraint constraint,
+      @NonNull INodeItem node,
+      @NonNull INodeItem target,
+      @NonNull String message,
+      @NonNull DynamicContext dynamicContext) {
     Level level = constraint.getLevel();
     if (isLogged(level)) {
-      logConstraint(level, node, newMissingIndexViolationMessage(constraint, node, target, message), null);
+      logConstraint(
+          level,
+          node,
+          newMissingIndexViolationMessage(
+              constraint,
+              node,
+              target,
+              message,
+              dynamicContext),
+          null);
     }
   }
 
   @Override
-  public void handlePass(IConstraint constraint, INodeItem node, INodeItem target) {
+  public void handlePass(
+      @NonNull IConstraint constraint,
+      @NonNull INodeItem node,
+      @NonNull INodeItem target,
+      @NonNull DynamicContext dynamicContext) {
     // do nothing
```

Comment:
Is this something where we _could_ allow DEBUG logging for constraints if we wanted to expose that later?

Reply:
I implemented this in the latest commit.